### PR TITLE
RES-1228: New and updated data utilities. 

### PR DESCRIPTION
**Pull request by mvacaporale**
This pull request includes additions to the data utilities:
* a dataloader for fake (i.e. random) data
* saver classes to aid in creating hdf5 files
* a class for creating temporary directories with fake data  (this was used to help write tests for the `imagenet_experiment` class which requires a saved dataset upon initialization)

As well, some minor enhancements were made to the original `dataset_utils.py` where the `HDF5Dataset` was modified to support non-image datatypes.

### CODE REVIEW

File: tests/unit/frameworks/pytorch/imagenet_experiment_test.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    def test_experiment_init(self):
+
+        exp = ImagenetExperiment()
+        temp_data = TempFakeSavedData(
+            train_size=10,
+            batch_size=2,
+            num_classes=10,
+        )
+
+        with temp_data as data:
```
Comment (subutai) [inline tests/unit/frameworks/pytorch/imagenet_experiment_test.py]: What is this testing? There's no assert here.

Comment (mvacaporale) [reply tests/unit/frameworks/pytorch/imagenet_experiment_test.py]: It's a simple attempt to instantiate the experiment with the temp-data. I can change the name to reflect that.

Comment (subutai) [reply tests/unit/frameworks/pytorch/imagenet_experiment_test.py]: My point is more about the test structure. How would this test fail? There's no assert to check whether it succeeded or not.  For example, you could use assertRaises to see if the setup_experiment call raised an exception.

Comment (mvacaporale) [reply tests/unit/frameworks/pytorch/imagenet_experiment_test.py]: Fair point. I'll add some assert statements for the data loaded by the experiment as that's what I hope to validate here. Please see the last commit (`ee341a6`).

**Comment by mvacaporale** [comment]
@lscheinkman Could you please review the changes I made to the `HDF5Dataset` class?

File: nupic/research/frameworks/pytorch/dataset_utils/dataset_utils.py [hunk 2]
```diff
@@ -350,11 +361,15 @@ class HDF5Dataset(VisionDataset):
         Other argument passed to :class:`VisionDataset` constructor
     """
 
-    def __init__(self, hdf5_file, root, num_classes=None, classes=None, **kwargs):
+    def __init__(
+        self, hdf5_file, root,
+        num_classes=None, classes=None, load_as_images=True, **kwargs
```
Comment (lscheinkman) [inline nupic/research/frameworks/pytorch/dataset_utils/dataset_utils.py]: Could you add a docstring for "load_as_images"

File: nupic/research/frameworks/pytorch/dataset_utils/dataset_utils.py [hunk 3]
```diff
@@ -378,7 +393,11 @@ def __init__(self, hdf5_file, root, num_classes=None, classes=None, **kwargs):
                 self._images = []
                 for class_name in self._classes:
                     group = hdf5_root[class_name]
-                    files = filter(is_image_file, group)
+
+                    if self._load_as_images:
+                        files = filter(is_image_file, group)
+                    else:
+                        files = group
```
Comment (lscheinkman) [inline nupic/research/frameworks/pytorch/dataset_utils/dataset_utils.py]: Why do you need this condition?

File: nupic/research/frameworks/pytorch/dataset_utils/fake_data.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+from torch.utils.data import DataLoader
+from torchvision.datasets import FakeData
+from torchvision.transforms import ToTensor
+
+__all__ = [
+    "FakeDataLoader",
+]
+
+
+class FakeDataLoader(DataLoader):
```
Comment (lscheinkman) [inline nupic/research/frameworks/pytorch/dataset_utils/fake_data.py]: it seems unnecessary and specific for imagenet (image_size=3, 224, 224). If anything it looks more like a test utility than a dataset utility

File: nupic/research/frameworks/pytorch/dataset_utils/hdf5_utils.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+        :param group_name: top level group name ("train", "val", etc)
+        :param lock: Lock object used to control write access to hdf5 file
+        :param image_path: Path object for the image file
+        """
+
+        wnid = class_name
+        if lock is not None:
+            lock.acquire()
+        hdf5_file = h5py.File(name=data_path, mode="a")
```
Comment (lscheinkman) [inline nupic/research/frameworks/pytorch/dataset_utils/hdf5_utils.py]: Could you add the "mode" as a parameter with default "a"?
Sometimes you may want to recreate the hdf5 file from scratch and pass "w"  instead.

**Comment by mvacaporale** [comment]
@lscheinkman I've made your suggested changes and have merged this PR.

**Pull request merged**